File: 45. Jump Game II/jump2.kt
import java.util.Scanner
import kotlin.math.*

class Solution {
    fun jump(nums: IntArray): Int {
        val n = nums.size
        var jumps = 0
        var end = 0
        var farthest = 0

        for (i in 0 until n - 1) {
            farthest = max(farthest, i + nums[i])

            if (i == end) {        // finished current jump range
                jumps++
                end = farthest
            }
        }
        return jumps
    }
}

fun main() {
    // val scanner=Scanner(System.`in`)
    // val n=scanner.nextInt()

    val nstr = readlnOrNull() ?: return

    val n = nstr.trim().toInt()

    // val nums =IntArray(n)
    // for(i in 0 until n){
    //     nums[i]=scanner.nextInt()
    // }
    val nums = readln().trim().split("\\s+".toRegex()).map { it.toInt() }.toIntArray()

    val obj = Solution()
    val ans = obj.jump(nums)

    println(ans)
}

File: 45. Jump Game II/jump2.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        int jumps = 0;
        int end = 0;
        int farthest = 0;

        for (int i = 0; i < n - 1; i++) {
            farthest = max(farthest, i + nums[i]);

            // time to make a jump
            if (i == end) {
                jumps++;
                end = farthest;
            }
        }
        return jumps;
    }
};


int main (){
    int n;
    cin>>n;

    vector<int> nums(n);
    for(int &i: nums){
        cin>> i;
    }

    Solution obj;
    int ans = obj.jump(nums);

    cout<<ans;

}
